go/src/koding/db/models/error.go

```diff
@@ -0,0 +1,65 @@
+package models
+
+import (
+	"fmt"
+	"strings"
+
+	mgo "gopkg.in/mgo.v2"
+)
+
+// NotFoundError is returned when requested operation on a resource
+// cannot be completed due to resource being missing. Most likely
+// this kind of errors means the resource was deleted outside
+// (shared ownership) and this error is returned to the caller,
+// so it recover if possible.
+//
+// TODO(rjeczalik): the db/mongo helpers should be decorating errors
+// instead, so database handling layer is transparent and no
+// library specific errors (e.g. mgo.ErrNotFound, pq.Error) are leaked.
+type NotFoundError struct {
+	Resource string // type / name of resource missing
+	Err      error  // underlying error, if any
+}
+
+// Error implements the builtin error interface.
+func (e *NotFoundError) Error() string {
+	return fmt.Sprintf("the %q resource was not found: %s", e.Resource, e.Err)
+}
+
+// Underlying gives the error value responsible for failure.
+func (e *NotFoundError) Underlying() error {
+	return e.Err
+}
+
+// ResError is a helper function that decorates the given err with more
+// meaningful error value, giving the caller context to recover.
+func ResError(err error, resource string) error {
```

Comment:
I'm not sure what should change here.

The `ResError` is something like `NewNotFoundErrorIf`, it wraps the err only under given condition. I think \*Error types do not require any special treatment to make sure they are not misused (unexporting fields etc.) - if wrong value is created (e.g. `var err error = &NotFoundError{Err: nil}`) it's treated as programmer error (I tried to follow the assumption stdlib makes for it's own error types - e.g. [assumption that for any OpError values Err field is not-nil](https://github.com/golang/go/blob/dd307da10c3582c195928d9bf073d1b0b01f2135/src/net/net.go#L448)).